neural_speed/convert/convert_quantized_gptj.py

```diff
@@ -0,0 +1,205 @@
+    args = parser.parse_args(args_in)
+
+    out_path = args.outfile.as_posix()
+    model_path = args.model.as_posix()
+
+    model, config, quantize_config = load_quantized_model(model_path)
+
+    tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
+    fout = open(out_path, "wb")
+
+    # 1. write hparams
+    hparams = config
+    n_layer = hparams["n_layer"]
+    fout.write(b"ggjt"[::-1])  #0x67676d6c)) # magic: ggml in hex
+    values = [
+        1,  # file version
+        hparams["vocab_size"],
+        hparams["n_embd"],
+        hparams["n_embd"] // hparams["n_head"],
+        hparams["n_head"],
+        hparams.get("n_head_kv", 0),  # multi-query attention
+        hparams["n_layer"],
+        hparams["rotary_dim"],
+        0
+    ]
+    fout.write(struct.pack("i" * len(values), *values))
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("f", 0))
+    fout.write(struct.pack("f", 0))
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))  # word_embed_proj_dim (for opt)
+    fout.write(struct.pack("i", 0))  # do_layer_norm_before (for opt)
+
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("f", hparams.get("rms_norm_eps", 1e-6)))  # rms norm eps
+    fout.write(struct.pack("f", 10000.0))  # freq_base
+    fout.write(struct.pack("f", 1.0))  # rope_factor
+    
+    fout.write(struct.pack("i", tokenizer.bos_token_id if tokenizer.bos_token_id is not None else 1))
+    fout.write(struct.pack("i", tokenizer.eos_token_id if tokenizer.eos_token_id is not None else 2))
+    fout.write(struct.pack("i", tokenizer.pad_token_id if tokenizer.pad_token_id is not None else -1))
+    fout.write(struct.pack("i", tokenizer.sep_token_id if tokenizer.sep_token_id is not None else -1))
+
+    # 2. vocab
+    byte_encoder = bytes_to_unicode()
+    byte_decoder = {v: k for k, v in byte_encoder.items()}
+
+    encoder = tokenizer.vocab
+    # Add added_tokens (special tokens) to the encoder
+    encoder_added = tokenizer.get_added_vocab()
+
+    for i, key in enumerate(sorted(encoder, key=encoder.get)):
+        # for key in encoder:
+        text = bytearray([byte_decoder[c] for c in key])
+        fout.write(struct.pack("i", len(text)))
+        fout.write(text)
+        if key not in encoder_added:
+            fout.write(struct.pack("f", 0.0 - i))
+        else:
+            fout.write(struct.pack("f", -10000))
+
+    # 3. write tensors
+    list_vars = model
+
+    convert_to_fp32_tensor("transformer.wte.weight", "transformer.wte.weight", list_vars, fout)
+    convert_to_fp32_tensor("transformer.ln_f.weight", "transformer.ln_f.weight", list_vars, fout)
+    convert_to_fp32_tensor("transformer.ln_f.bias", "transformer.ln_f.bias", list_vars, fout)
+    convert_to_fp32_tensor("lm_head.bias", "lm_head.bias", list_vars, fout)
+    convert_to_fp32_tensor("lm_head.weight", "lm_head.weight", list_vars, fout)
```

Comment:
could we detect whether lm_head.weight need to convert to qx or just keep fp32? 
i guess we have the chance, e.g. check whether lm_head.weight have qzeros/scales tensor member.
this may helpful if inc can quantize lm_head to qx in the future.

Reply:
That's good idea. I'll enable this feature when quantized lm_head.weight is provied.